src/classes/BDI_DataImportService.cls

```diff
@@ -0,0 +1,2252 @@
+            if (strPreferred != null)
+                return null;
+            strPreferred = 'Mobile';
+        }
+        if (di.get(strCx + '_Other_Phone__c') != null) {
+            if (strPreferred != null)
+                return null;
+            strPreferred = 'Other';
+        }
+        return strPreferred;
+    }
+
+    /*******************************************************************************************************
+    Key Matching Explained
+
+    First, we create a key which is the Salesforce Id of the Contact.
+    Second, if Contact Unique Id is specified, we create a key with that unique id.
+    Then we use a compound key of Contact Firstname, Lastname, Email, and Phone (with a | between each) to
+    match against existing Contacts,
+    and also those Contacts we are creating. We call this compound key a diKey. Our mapDIKeyToContact
+    is used to lookup a given Contact from a diKey. In order to support settings that tell us which fields
+    to use in the match, a given Contact can have multiple keys. Eg., if the setting is Email Only, and
+    we have an existing Contact of Fred Smith fred@smith.com, then we need to support matching the following
+    diKeys:
+        Fred|Smith|fred@smith.com
+        Fred||fred@smith.com
+        |Smith|fred@smith.com
+        ||fred@smith.com
+
+    In this same example, if our DI record is John Doe john@doe.com, then its diKeys would be:
+        John|Doe|john@doe.com
+        John||john@doe.com
+        |Doe|john@doe.com
+        ||john@doe.com
+
+    But if another DI record is just Bob Dole (no email), then we know it can't match an existing Contact
+    since it has no email, but we still need to create a unique diKey for this record so that we can track
+    its Contact, as well as map other DI records with the same name to the same contact.  Thus we'll create
+    the diKey:
+        Bob|Dole|
+
+    So the rules can be explained as:
+        Existing contacts only add the specific keys the rules can match against.
+        Di's add the specific keys the rules can match against, and if none created, a single unique key
+
+    ************/
+
+    /*******************************************************************************************************
+    * @description return all the diKeys for a given contact
+    * @param con the Contact
+    * @return List<string> a list of diKeys
+    */
+    private list<string> listDiKeyContact(Contact con) {
+        if (con == null)
+            return null;
+
+        list<string> listDiKey = new list<string>();
+
+        // we need any Contact SFId key to always be checked first when matching
+        if (con.Id != null) {
+            listDiKey.add(con.Id);
+        }
+
+        // we need the Contact CustomId key to always be checked second when matching
+        string strUniqueId = '';
+        if (isCustomIDInContactMatchRules && (strUniqueId = strNull(con.get(diSettings.Contact_Custom_Unique_ID__c))) != '') {
```

Comment:
This is confusing to me. It looks like strUniqueId is getting a Boolean value.

Reply:
I see you're doing the same thing in a bunch of places. I guess it's fine ;)

Reply:
sorry it isn't clear.  it's assigning the unique Id to the variable (so it can be used again later), and then comparing to ''.